### PR TITLE
CR-2791: [Not CAR] [GCP] Ensure that instances are not configured to use the default service account

### PR DESCRIPTION
**Pull request by Imanuel-Mizrahi**

### CODE REVIEW

File: tests/knowledge/context/gcp/test_compute_instance.py [hunk 1]
```diff
@@ -73,3 +73,24 @@ def test_shielded_vm_enable_vtpm_not_integrity(self, ctx: GcpEnvironmentContext)
         self.assertTrue(compute.shielded_instance_config.enable_secure_boot)
         self.assertFalse(compute.shielded_instance_config.enable_integrity_monitoring)
         self.assertTrue(compute.shielded_instance_config.enable_vtpm)
+
+    @context(module_path="default_service_account")
+    def test_default_service_account(self, ctx: GcpEnvironmentContext):
+        self.assertTrue(len(ctx.compute_instances), 2)
```
Comment (Tomerva) [inline tests/knowledge/context/gcp/test_compute_instance.py:79]: why are there 2 instances but only assertion on one of them?

File: tests/knowledge/rules/gcp/non_context_aware/test_compute_instance_do_not_use_default_service_account_rule.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+from cloudrail.knowledge.rules.gcp.non_context_aware.compute_instance_do_not_use_default_service_account_rule import ComputeInstanceDoNotUseDefaultServiceAccountRule
+
+
+class TestComputeInstanceDoNotUseDefaultServiceAccountRule(TestCase):
+    def setUp(self):
+        self.rule = ComputeInstanceDoNotUseDefaultServiceAccountRule()
+
+    @parameterized.expand(
+        [
+            ["default_service_account_used", '37924132841-compute@developer.gserviceaccount.com', True,],
```
Comment (Tomerva) [inline tests/knowledge/rules/gcp/non_context_aware/test_compute_instance_do_not_use_default_service_account_rule.py:17]: redundant comma at the end

**Review by Tomerva — CHANGES_REQUESTED** [review]

File: cloudrail/knowledge/rules/gcp/non_context_aware/compute_instance_do_not_use_default_service_account_rule.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+    def execute(self, env_context: GcpEnvironmentContext, parameters: Dict[ParameterType, any]) -> List[Issue]:
+        issues: List[Issue] = []
+        for compute_instance in env_context.compute_instances:
+            if compute_instance.service_account \
+                and not compute_instance.service_account.email or compute_instance.service_account.email.split('-')[0].isnumeric():
+                issues.append(
+                    Issue(
+                        f"The {compute_instance.get_type()} `{compute_instance.get_friendly_name()}` has service_account block "
+                        f"attributes email uses default service account [project_number]-compute@developer.gserviceaccount.com",
```
[the remark this answers is not in the document]
Comment (Imanuel-Mizrahi) [reply cloudrail/knowledge/rules/gcp/non_context_aware/compute_instance_do_not_use_default_service_account_rule.py:21]: @Tomerva , is that better:
f"The {compute_instance.get_type()} `{compute_instance.get_friendly_name()}` has service_account block "
                        f"email attribute which uses default service account [project_number]-compute@developer.gserviceaccount.com",
                        ?

File: tests/knowledge/context/gcp/test_compute_instance.py (hunk 1 above)
Comment (Imanuel-Mizrahi) [reply tests/knowledge/context/gcp/test_compute_instance.py:79]: Well, the second is pretty much the same, just the name differs.
I can do it if you would like

Comment (Imanuel-Mizrahi) [reply tests/knowledge/context/gcp/test_compute_instance.py:82]: @Tomerva , changed that into a "for" loop, to run on both

File: cloudrail/knowledge/context/base_context_builders/base_terraform_builder.py [hunk 4]
```diff
@@ -55,8 +55,11 @@ def _finalize_component(instance: Mergeable, iac_state: IacState, attributes: di
         instance.iac_state = iac_state
         TerraformResourceFinder.add_resource(instance)
         for tag_key in ('tags_all', 'tags', 'tag'):
-            if instance.is_tagable and attributes and isinstance(attributes.get(tag_key), dict):
-                instance.tags = attributes[tag_key] or {}
+            if instance.is_tagable and attributes and attributes.get(tag_key):
+                if isinstance(attributes.get(tag_key), dict):
+                    instance.tags = attributes[tag_key] or {}
+                if isinstance(attributes.get(tag_key), list):
```
Comment (Tomerva) [inline cloudrail/knowledge/context/base_context_builders/base_terraform_builder.py:61]: change to elif

File: cloudrail/knowledge/utils/tags_utils.py [hunk 5]
```diff
@@ -46,9 +57,15 @@ def _hash_tags_dict(dictionary: dict, salt: str) -> dict:
 
     return result
 
+def _hash_gcp_tags_dict(dictionary: dict, salt: str) -> dict:
```
Comment (Tomerva) [inline cloudrail/knowledge/utils/tags_utils.py:60]: rename to _hash_gcp_labels_dict

File: cloudrail/knowledge/utils/tags_utils.py [hunk 6]
```diff
@@ -46,9 +57,15 @@ def _hash_tags_dict(dictionary: dict, salt: str) -> dict:
...
 
-def filter_tags(tags: dict):
-    excluded_tags = ['aws:cloudformation:stack-name',
-                     'aws:cloudformation:logical-id',
-                     'aws:cloudformation:stack-id']
-    return {k: v for k, v in tags.items() if k not in excluded_tags}
+    return _hash_tags(dictionary, salt, allowed_tags)
+
+def get_gcp_labels(raw_gcp_data: dict) -> Optional[dict]:
+    gcp_raw_tags = raw_gcp_data.get('labels')
```
Comment (Tomerva) [inline cloudrail/knowledge/utils/tags_utils.py:67]: rename to gcp_raw_labels

**Review by Tomerva — APPROVED** [review]

**Pull request merged**